docs/specs/saml.md

```diff
@@ -0,0 +1,697 @@
+# SAML
+
+- [Web Browser SSO](#1)
+  - [SP Initiated](#1_1)
+  - [IdP Initiated](#1_2)
+  - [Configs](#1_3)
+  - [Supported Parameters in `<AuthnRequest>`](#1_4)
+  - [Support Pre-Selecting User by SAML Subject in Authentication Flow](#1_5)
+  - [The Login Endpoint](#1_6)
+- [Single Logout](#2)
+  - [Configs](#2_1)
+  - [Supported Parameters in `<LogoutRequest>`](#2_2)
+  - [The Logout Endpoint](#2_3)
+- [NameID](#3)
+  - [Supported NameIDFormat](#3_1)
+    - [Configure NameID when using urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified](#3_1_1)
+  - [Configs](#3_2)
+  - [Unsupported `NameIDFormat`](#3_3)
+- [Attributes](#4)
+  - [Customizing the attributes](#4_1)
+  - [Use a hook to compute SAML attributes](#4_2)
+    - [Payload](#4_2_1)
+    - [Response](#4_2_2)
+- [Bindings](#5)
+  - [Supported Bindings](#5_1)
+  - [Unsupported Bindings](#5_2)
+- [Signing](#6)
+  - [Configs](#6_1)
+  - [Secrets](#6_2)
+- [Metadata](#7)
+- [Debug Logging](#8)
+  - [Configs](#8_1)
+- [Profiles not supported at the moment](#9)
+- [An example with all configurable options](#10)
+- [Service Provider Support](#11)
+
+## <a id="1"></a> Web Browser SSO
+
+### <a id="1_1"></a> SP Initiated
+
+Ref: [SAMLProfile](https://groups.oasis-open.org/higherlogic/ws/public/download/35389/sstc-saml-profiles-errata-2.0-wd-06-diff.pdf) Section 4.1
+
+```mermaid
+sequenceDiagram
+    participant ua as User Agent (Alice)
+    participant sp as SP (Web App) <br/> https://example.com
+    participant ag as IdP (Authgear) <br/> https://example.authgear.cloud
+    ua->>+sp: [1] Request for protected resources https://example.com/flights
+    sp->>-ua: [2] Redirect to https://example.authgear.cloud/saml2/login/ENTITY_ID with <AuthnRequest>
+    ua->>+ag: [3] Send <AuthnRequest> to https://example.authgear.cloud/saml2/login
+    Note right of ag: Alice enters user credentials
+    ag->>-ua: [4] Redirect to https://example.com/acs with <Response>
+    ua->>+sp: [5] Send <Response> to https://example.com/acs
+    Note right of sp: Authenticated as Alice
+    sp->>-ua: [6] Response with protected resources https://example.com/flights
+
+```
+
+- The login endpoint `https://example.authgear.cloud/saml2/login/ENTITY_ID` is from the metadata.
```

Comment:
It seems that we do not have a dedicated section to specify the metadata endpoint.



Another question: How does other authentication software name these endpoints I mean `https://example.authgear.cloud/saml2/login/ENTITY_ID` vs `https://example.authgear.cloud/saml2/service_providers/ENTITY_ID/login`

Reply:
Another question: Maybe we should rename the term `Entity ID` to `Service Provider ID` in this spec. Or the SAML spec uses the term `Entity ID` so we want to be consistent with the SAML spec?

Reply:
It is here https://github.com/tung2744/authgear-server/blob/design-saml/docs/specs/saml.md#7

Maybe I can document the endpoint too.

 > How does other authentication software name these endpoints

For auth0, it is like https://dev-y17ryy2xy2nvgpk1.us.auth0.com/samlp/lhImeA5MPX4nthJGxb5MCDW5AeK6QEYL

> Maybe we should rename the term Entity ID to Service Provider ID in this spec. Or the SAML spec uses the term Entity ID so we want to be consistent with the SAML spec?


Yes. Entity ID is a term used in the spec. And it will also appear in the metadata. Every "entity", such as the IdP and Service Provider, will have its own entity id.
However, the spec does not use the entity id in any url.